godot-core/src/builtin/vector_macros.rs

```diff
@@ -0,0 +1,221 @@
+        impl std::ops::$operator<$scalar> for $vector {
+            fn $func(&mut self, rhs: $scalar) {
+                $(
+                    self.$components.$func(rhs);
+                )*
+            }
+        }
+    }
+}
+
+/// Implements all common arithmetic operators on a built-in vector type.
+macro_rules! impl_vector_operators {
+    (
+        // Name of the vector type to be implemented, for example `Vector2`.
+        $vector:ident,
+        // Type of each individual component, for example `f32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*)$(,)?
+    ) => {
+        impl_vector_unary_operator!($vector, $scalar, ($($components),*), Neg, neg);
+        impl_vector_binary_operator!($vector, $scalar, ($($components),*), Add, add);
+        impl_vector_binary_operator!($vector, $scalar, ($($components),*), Sub, sub);
+        impl_vector_binary_operator!($vector, $scalar, ($($components),*), Mul, mul);
+        impl_vector_binary_operator!($vector, $scalar, ($($components),*), Div, div);
+        impl_vector_binary_operator!($vector, $scalar, ($($components),*), Rem, rem);
+        impl_vector_assign_operator!($vector, $scalar, ($($components),*), AddAssign, add_assign);
+        impl_vector_assign_operator!($vector, $scalar, ($($components),*), SubAssign, sub_assign);
+        impl_vector_assign_operator!($vector, $scalar, ($($components),*), MulAssign, mul_assign);
+        impl_vector_assign_operator!($vector, $scalar, ($($components),*), DivAssign, div_assign);
+        impl_vector_assign_operator!($vector, $scalar, ($($components),*), RemAssign, rem_assign);
+    }
+}
+
+/// Implements `Index` and `IndexMut` for a vector type, using an enum to indicate the desired axis.
+macro_rules! impl_vector_index {
+    (
+        // Name of the vector type to be implemented, for example `Vector2`.
+        $vector:ident,
+        // Type of each individual component, for example `f32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the enum type for the axes, for example `Vector2Axis`.
+        $axis_enum:ident,
+        // Names of the enum variants, with parenthes, for example `(X, Y)`.
+        ($($axis_variants:ident),*)$(,)?
+    ) => {
+        impl std::ops::Index<$axis_enum> for $vector {
+            type Output = $scalar;
+            fn index(&self, axis: $axis_enum) -> &$scalar {
+                match axis {
+                    $($axis_enum::$axis_variants => &self.$components),*
+                }
+            }
+        }
+
+        impl std::ops::IndexMut<$axis_enum> for $vector {
+            fn index_mut(&mut self, axis: $axis_enum) -> &mut $scalar {
+                match axis {
+                    $($axis_enum::$axis_variants => &mut self.$components),*
+                }
+            }
+        }
+    }
+}
```

Comment:
That's a nice idea! 🙂

Also here: `$Vector:ty`, `$Scalar:ty`, `$AxisEnum:ty` (PascalCase).
They can all be `ty`, no?

Reply:
Yes. Also wrote `$AxisVariants:ident` for consistency with casing of enum variants (in Rust, that is...).